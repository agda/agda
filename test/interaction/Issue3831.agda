open import Agda.Builtin.List
open import Agda.Builtin.Unit
open import Agda.Builtin.Nat
open import Agda.Builtin.String
  renaming (primShowNat to show)
open import Agda.Builtin.Reflection
  renaming (bindTC to _>>=_; returnTC to return)

pattern vArg t = arg (arg-info visible relevant) t
pattern var₀ x = var x []

infixr 10 _++_
_++_ = primStringAppend

postulate
  whatever : ∀ {a} {A : Set a} → A

macro
  test1 : Nat → Term → TC _
  test1 n _ =
    extendContext (vArg (quoteTerm Nat)) do
      var₀ i ← quoteTC n where _ → whatever
      m ← unquoteTC {A = Nat} (var₀ 0)
      var₀ j ← quoteTC m where _ → whatever
      extendContext (vArg (quoteTerm Nat)) do
        var₀ k ← quoteTC n where _ → whatever
        var₀ l ← quoteTC m where _ → whatever
        typeError (strErr (show i ++ show k ++ show j ++ show l) ∷ [])

  test2 : Term → TC _
  test2 hole = do
    st ← quoteTC Set
    t ← extendContext (vArg st) do
      v ← unquoteTC {A = Set} (var₀ zero)
      extendContext (vArg (var₀ zero)) do
        _ ← unquoteTC {A = v} (var₀ zero)
        return tt
    u ← quoteTC t
    unify hole u

  test3 : Nat → Term → TC _
  test3 n _ = do
    m      ← extendContext (vArg (quoteTerm Nat)) (return n)
    var₀ i ← quoteTC m where _ → whatever
    typeError (strErr (show i) ∷ [])

  localvar : Term → TC _
  localvar _ = do
    m ← extendContext (vArg (quoteTerm Nat)) (unquoteTC {A = Nat} (var₀ 0))
    typeError (strErr (show m) ∷ [])
