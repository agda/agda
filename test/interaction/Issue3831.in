top_command (cmd_load currentFile [])
top_command (cmd_infer_toplevel Normalised "test2")
top_command (cmd_infer_toplevel Normalised "\\ n -> test1 n")
top_command (cmd_infer_toplevel Normalised "\\ n m l -> test1 n")
top_command (cmd_infer_toplevel Normalised "\\ n -> test3 n")
top_command (cmd_infer_toplevel Normalised "\\ n m -> test3 n")
top_command (cmd_infer_toplevel Normalised "localvar")
