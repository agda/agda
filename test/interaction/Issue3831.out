(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-status-action "Checked")
(agda2-info-action "*All Done*" "" nil)
((last . 1) . (agda2-goals-action '()))
(agda2-status-action "Checked")
(agda2-info-action "*Inferred Type*" "‚ä§" nil)
(agda2-info-action "*Error*" "1,8-15 1201 when checking that the expression unquote (test1 n) has type _130" nil)
(agda2-highlight-load-and-delete-action)
(agda2-status-action "")
(agda2-info-action "*Error*" "1,12-19 3401 when checking that the expression unquote (test1 n) has type _138" nil)
(agda2-highlight-load-and-delete-action)
(agda2-status-action "")
(agda2-info-action "*Error*" "1,8-15 0 when checking that the expression unquote (test3 n) has type _130" nil)
(agda2-highlight-load-and-delete-action)
(agda2-status-action "")
(agda2-info-action "*Error*" "1,10-17 1 when checking that the expression unquote (test3 n) has type _134" nil)
(agda2-highlight-load-and-delete-action)
(agda2-status-action "")
(agda2-info-action "*Error*" "1,1-9 Local variable 'x' escaping in result of extendContext: x when checking that the expression unquote localvar has type _126" nil)
(agda2-highlight-load-and-delete-action)
(agda2-status-action "")
